Sprint 2/step 2

eglazkov opened the pull request.

eglazkov committed:
подготовка хранилища и уход от reactContext к redux

eglazkov committed:
доработка функционала переключения табов при скролле

eglazkov committed:
вынес элемент конструктора в отдельный компонент

eglazkov committed:
d'n'd сортировка ингредиентов в конструкторе

eglazkov committed:
реализован механизм перетаскивания d'n'd ингредиентов в конструктор бургера

eglazkov committed:
исправление ошибки замены ингредиентов bun

eglazkov committed:
немного доработал поведение и стилизацю dropTrget

irinamozes commented on src/services/ingredients/action-creators.js:
```diff
@@ -0,0 +1,48 @@
+};
+
+export const increaseCountAction = (id) => dispatch => {
+  dispatch({type: INCREASE_INGREDIENT_COUNT, payload: {id}});  
```
В этом файле все акции, кроме fetchIngredientsAction, синхронные. Креэйторы синхронных акций просто возвращают объекты акций
https://praktikum.yandex.ru/learn/react/courses/61ff90eb-9909-4a41-b910-970ae7a90b38/sprints/6023/topics/951327ed-1414-4e31-89d0-70ed1e048bd9/lessons/4aaf6377-2359-486f-af2f-7ba323c8aaf5/,
без посылки их в методах dispatch, то есть синхронные акции должны быть просто объектами. Думаю, это бы оптимизировало работу thunk, суть которого я попыталась сформулировать в общем комментарии.

irinamozes commented:
Проект очень хороший.

Правильно используется усилитель thunk https://praktikum.yandex.ru/learn/react/courses/61ff90eb-9909-4a41-b910-970ae7a90b38/sprints/6023/topics/951327ed-1414-4e31-89d0-70ed1e048bd9/lessons/6ec202d8-c61c-4a1b-ad27-abd2e658612b/.

Его задача - получить акцию, посланную с помощью dispatch, и проанализировать, функция это, или  объект. Если функция, то эта функция выполняется и при выполнении посылает акции-объекты. А с акциями-объектами thunk поступает так: либо передаёт их следующему усилителю, либо передаёт на обработку редьюсеру, если усилителей больше нет.

Но, есть несколько замечаний.

Что нужно исправить.

1. При сортировке массива ингредиентов, Вы ингредиенты переставляете местами. Но. сортировка- это изменение порядка следования элементов, а не перестановка их местами.
Думаю, сортировка должна происходить по следующему алгоритму:
если пользователь хочет какой-то ингредиент поставить на место другого ингредиента, он должен перемещаемый ингредиент совместить с целевым ингредиентом. После этого, другие ингредиенты, смещаются вниз на один слой (если перемещаемый ингредиент двигался с низу), чтобы заполнить место перемещаемого ингредиента, а перемещаемый ингредиент встаёт на освободившееся место целевого ингредиента. Точно также ингредиенты смещаются вверх, если перемещаемый ингредиент был расположен сверху по отношению к целевому.

Что можно лучше.

1. Лучше преобразовать креэйторы синхронных акций.
https://praktikum.yandex.ru/learn/react/courses/61ff90eb-9909-4a41-b910-970ae7a90b38/sprints/6023/topics/951327ed-1414-4e31-89d0-70ed1e048bd9/lessons/4aaf6377-2359-486f-af2f-7ba323c8aaf5/
(подробный комментарий в ingredients/action-creators.js)

2. Лучше использовать хук useSelector для передачи данных из store в пропсы компонентов.
https://praktikum.yandex.ru/learn/react/courses/61ff90eb-9909-4a41-b910-970ae7a90b38/sprints/6023/topics/951327ed-1414-4e31-89d0-70ed1e048bd9/lessons/19f936c8-1df6-4f55-85ac-d68cd34b78ff/

3. Можно не создавать креэйторы синхронных акций, а использовать акции как объекты.

eglazkov committed:
Исправления:
1. Сортировка массива ингредиентов в конструкторе - готово
Улучшения:
1. Преобразование крейтеров синхронных акций - готово
2. Использование хука useSelector - уже использовался ранее (services/counstructor/index, services/ingredients/index, services/order/index) - в результате вызова кастомного хука, который возвращает данные из стора и крейторы акций
3. Отказ от крейтеров синхронных акций - решено оставить для выполнения принципа DRY

eglazkov committed:
удалил признак верх с начинки

eglazkov committed:
сброс общей суммы заказа после оформления заказа

eglazkov committed:
Merge pull request #9 from eglazkov/sprint-2/step-2-fix-2

сброс общей суммы заказа после оформления заказа

eglazkov committed:
Merge pull request #8 from eglazkov/sprint-2/step-2-fix

Исправления:
